src/couch/src/couch_httpd_auth.erl

```diff
@@ -695,3 +700,33 @@ authentication_warning(#httpd{mochi_req = Req}, User) ->
         "~p: Authentication failed for user ~s from ~s",
         [?MODULE, User, Peer]
     ).
+
+verify_hash_names(HashAlgorithms, SupportedHashFun) ->
+    verify_hash_names(HashAlgorithms, SupportedHashFun, []).
+verify_hash_names([], _, HashNames) ->
+    HashNames;
+verify_hash_names([H | T], SupportedHashFun, HashNames) ->
+    try
+        HashAtom = binary_to_existing_atom(H),
+        Result =
+            case lists:member(HashAtom, SupportedHashFun) of
+                true -> HashNames ++ [HashAtom];
```

Comment:
while it's unlikely a problem here, with very short lists, we try to avoid `++`. since order doesn't matter I think `[HashAtom | HashNames]` is preferable.

Reply:
Okay, but the order of the hash algos is important. Assume we have
```ini
hash_algorithms = sha256, sha
```
then we get as result
```erlang
[sha, sha256]
```
and the cookie is generated with the first algorithm.
Maybe your solution 
```erlang
true -> [HashAtom | HashNames]
```
in combination with
```
verify_hash_names([], _, HashNames) ->
    lists:reverse(HashNames);
```

Is this better?

Reply:
maybe combine these things to avoid the problem entirely. iterate through the `re:split` list of hash algorithm names in order, moving to the next if its not a valid algorithm name or if the hash doesn't match for valid names, and then return a failure if the list is empty.

Reply:
I added both combinations and I think it's fine...